Feature: Structured logging support (JSON)

SadiHassan opened the pull request:
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

This feature will provide JSON formatted log output in console. The purpose is to allow easier ingestion and query in log analysis tools.

1. A sample log output might look like this.
```{"datetime":"2022.03.28 12:17:54.827035","thread_id":"790151","query_id":"2441139","priority":"Debug","source":"KeeperTCPHandler","message":"Received close event with xid 2147483647 for session id #7"}```
2. User will be able to turn on/off `JSON` output.
For example consider the sample tag below under ```<logger>``` (that will be written in `config.xml`). 
```
<format>
     <json>true</json>
</format>
```
If user deletes or comments out the `format` or `json` tag, usual log output will prevail. 

Note: This is the first stage of implementation of #35366. As @blinkov suggested, once this part gets accepted, the other part can be implemented right after.
...


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

SadiHassan committed:
Feature: Structured logging support(JSON)

SadiHassan commented:
@nikitamikhaylov I see stylecheck failed for the PR description template modified by me. Did you add "Can be tested" for your later manual test? (I mean, manually run other tests that checks coding and functionality?)

SadiHassan committed:
structured logging - commit to make PR description in correct format

SadiHassan committed:
structured logging - RE:commit to make PR description in correct format

SadiHassan committed:
structured logging - style correction

SadiHassan committed:
structured logging - style (whitespace) correction

SadiHassan committed:
structured logging - style + others suspected area corrections

SadiHassan committed:
structured logging - style - removed space between if and first bracket

alexey-milovidov commented on base/loggers/Loggers.cpp:
```diff
@@ -49,7 +50,6 @@ void Loggers::buildLoggers(Poco::Util::AbstractConfiguration & config, Poco::Log
         if (auto log = text_log.lock())
             split->addTextLog(log, text_log_max_priority);
 #endif
-
```
This change looks irrelevant.

SadiHassan replied:
Agree. Will remove.

alexey-milovidov commented on base/loggers/Loggers.h:
```diff
@@ -10,15 +10,17 @@
+class AbstractConfiguration;
 }
 
+extern bool log_format_json;
```
Global variables makes code feel bad.

SadiHassan replied:
The problem here is, we need to get the config settings which is found in `Loggers.cpp`. We also need to use the `config` properties into `OwnFormattingChannel` and `OwnPatternFormatter`. But there is no data passing from `Loggers` to any of them that passes the `config` property. That's why global was used. However, I found another way that is adding a static member inside `ExtendedLogMessage` class and accessing that member from `Loggers.cpp`. This is slightly better than global namespace. This will keep it under `DB::ExtendedLogMessage`. Do you have any suggestion around this?

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+    writeCString("\"", wb);
+    writeCString(" [ ", wb);
+    if (color)
+        writeString(setColor(intHash64(msg_ext.thread_id)), wb);
```
JSON does not have colors.

SadiHassan replied:
Agreed. Will change it.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+
+    writeCString(",", wb);
+
+    writeCString("\"", wb);
```
writeChar can be used.

SadiHassan replied:
Agreed. Will change this.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+    writeCString("\"", wb);
+    writeCString(":", wb);
+    writeCString("\"", wb);
+    writeCString(" [ ", wb);
```
There is no sense to put decorated string ` [1]` instead of number.

SadiHassan replied:
Agreed. Will change this. When I found test case failing because of `assert` in python that checks specific format with `[` or `{`, I added those to make test pass.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+
+    if (color)
+        writeString(setColor(std::hash<std::string>()(msg_ext.query_id)), wb);
+    DB::writeString(msg_ext.query_id, wb);
```
This is wrong, because the string is not escaped.

SadiHassan replied:
Agreed. Will change this.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+
+    if (color)
+        writeString(setColor(std::hash<std::string>()(msg.getSource())), wb);
+    DB::writeString(msg.getSource(), wb);
```
The string is not escaped.

SadiHassan replied:
Agreed. Will change this.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+    writeCString(":", wb);
+
+    writeCString("\"", wb);
+    DB::writeString(msg.getText(), wb);
```
The string is not escaped.

SadiHassan replied:
Agreed. Will change this.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on base/loggers/OwnPatternFormatter.cpp:
```diff
@@ -7,13 +7,118 @@
+    DB::writeString(msg.getText(), wb);
+    writeCString("\"", wb);
+
+    writeCString("}", wb);
```
Missing file, line and other fields from system.text_log.

SadiHassan replied:
Agreed. Will add them. Is there any suggestion of their sequence?

alexey-milovidov replied:
The ordering can be identical to the `system.text_log` table.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on programs/server/config.xml:
```diff
@@ -60,6 +60,18 @@
+        You can specify log format (for now, JSON only). In that case, the console log will be printed
+        in JSON format.
+        For example, like this below:
+        {"datetime":"2022.03.28 12:17:54.827035","thread_id":"790151","query_id":"2441139","priority":"Debug","source":"KeeperTCPHandler","message":"Received close event with xid 2147483647 for session id #7"}  
```
DateTime should be in unix timestamp or in ISO-8601 with UTC.

SadiHassan replied:
Agreed. Will change this.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on programs/server/config.xml:
```diff
@@ -60,6 +60,18 @@
+        You can specify log format (for now, JSON only). In that case, the console log will be printed
+        in JSON format.
+        For example, like this below:
+        {"datetime":"2022.03.28 12:17:54.827035","thread_id":"790151","query_id":"2441139","priority":"Debug","source":"KeeperTCPHandler","message":"Received close event with xid 2147483647 for session id #7"}  
```
query_id does not typically looks like this.

SadiHassan replied:
Agreed. This was a made up thing to fill the explanation. Will update with originals.

SadiHassan replied:
change submitted in updated PR.

alexey-milovidov commented on programs/server/config.xml:
```diff
@@ -60,6 +60,18 @@
+        -->
+        <format>
+            <json>true</json>
+        </format>
```
Need to write a test.

alexey-milovidov replied:
For better consistency, let's make JSON formatting exactly as JSONEachRow formatting of system.text_log.

SadiHassan replied:
Any idea of what kind of test should it run? My guess is, if the `config` has `json` in it, the test program should grab log message and see if the message is in correct log format or not. Anything else? Also, can you hint any test file to understand how to write a test function for this?

alexey-milovidov replied:
I can be integration test in `tests/integration`.
It can check that server writes logs in JSON, that JSON is valid, and not necessarily anything else.

alexey-milovidov reviewed (changes requested):
.

alexey-milovidov commented:
Please address review comments.

alexey-milovidov commented:
Abandoned.

SadiHassan committed:
resubmit with changes (without testing)

SadiHassan committed:
resubmit with changes of config.xml (without testing)

larryluogit commented on base/loggers/OwnFormattingChannel.cpp:
```diff
@@ -1,18 +1,23 @@
 #include "OwnFormattingChannel.h"
+#include <iostream> //mallik--remove pls
```
^^

SadiHassan committed:
added inregration test and other fixes

SadiHassan committed:
stylecheck fix in Python file

SadiHassan commented:
Hi @alexey-milovidov, I suspect the failing test cases are not related to the changes I made. Can you please have a look and suggest?

SadiHassan committed:
Merge branch 'master' into structured-logging-support
To resolve merge conflict with master

SadiHassan committed:
Merge branch 'master' into structured-logging-support

SadiHassan committed:
added writeJSONString where applicable

SadiHassan commented:
@alexey-milovidov submitted the enhanced version of this feature in another PR #36896. Once this one is reviewed, you can take a look at that.

alexey-milovidov commented:
@SadiHassan 

Instead of using global variable, we can implement it in a different `PatternFormatter`, similar to `OwnPatternFormatter` (you can name it `OwnJSONPatternFormatter`) and create it in `BaseDaemon.cpp` according to configuration.

SadiHassan commented:
> @SadiHassan
> 
> Instead of using global variable, we can implement it in a different `PatternFormatter`, similar to `OwnPatternFormatter` (you can name it `OwnJSONPatternFormatter`) and create it in `BaseDaemon.cpp` according to configuration.
@alexey-milovidov 
Okay. In that case, we will have to make changes in the following way: According to configuration, we will have to add `OwnJSONPatternFormatter` in `BaseDaemon.cpp` (1 place, inside `setupWatchdog()` function) and in `Loggers.cpp` (3 places, under conditions of `logpath`, `errorlog_path` and `use_syslog`. Also, to make it work, we need and overloaded constructor of `OwnFormattingChannel` which will take `OwnJSONPatternFormatter` as a parameter (instead of `OwnPatternFormatter`). Finally, in `OwnFormattingChannel.cpp`, we have to call our `formatExtendedJSON()` function based on `pFormatterJSON` is not null. 
Please share your thoughts about this.

SadiHassan committed:
code restructured to avoid global variable

SadiHassan committed:
code restructured to avoid global variable

SadiHassan commented:
@alexey-milovidov as discussed, removed global variable and created new class `OwnJSONPatternFormatter` and applied necessary changes in files as explained above.

SadiHassan committed:
stylecheck correction in BaseDaemon.cpp

SadiHassan committed:
stylecheck correction in BaseDaemon.cpp

SadiHassan committed:
resolved merge conflict

kashwy committed:
Kusto-phase1: Add Support to Kusto Query Language

This is the initial implement of Kusto Query Language.

in this commit, we support the following features as MVP :

Tabular expression statements
Limit returned results
Select Column (basic project)
sort, order
Perform string equality operations
Filter using a list of elements
Filter using common string operations
Some string  operators
Aggregate by columns
Base aggregate functions
only support avg, count ,min, max, sum
Aggregate by time intervals

SadiHassan committed:
Kusto summarize init

SadiHassan committed:
added single unit test case for summarize bin()

SadiHassan committed:
removed unwanted comments

SadiHassan committed:
corrected unit test

kashwy committed:
Kusto-phase1 :
Add new test cases

kashwy committed:
Kusto-phase1: Fixed the bug for KQL filer with multiple operations

kashwy committed:
Kusto-phase1: Fixed style

kashwy committed:
Kusto-pahse1: Fixed moy style issues.

kashwy committed:
Kusto-phase1: Fixed misleading indentation

kashwy committed:
Kusto-pahse2: Add support for multiple summarize

kashwy committed:
Kusto-phase2 : Added KQL functions interface.
changed the summarize class for new aggregation functions

SadiHassan committed:
conflict resolution in Basedaemon.cpp

SadiHassan committed:
conflict resolution in Basedaemon.cpp

kashwy committed:
Kusto-phase2: Add KQL functions parser

kashwy committed:
Kusto-phase2: Add common function to get argument for function convertion

kashwy committed:
Kusto-phase2: add kusto_auto dialect

kashwy committed:
Kusto-phase2: Add alias support

bkuschel committed:
Create test.yml

bkuschel committed:
New Tests

bkuschel committed:
Update tests

SadiHassan committed:
Merge branch 'master' into structured-logging-support

SadiHassan committed:
fixed build error
